client/src/nonblocking/tpu_client.rs

```diff
@@ -0,0 +1,557 @@
+                    pubsub_client,
+                    exit,
+                )
+                .await
+            })
+        });
+
+        Ok(LeaderTpuService {
+            recent_slots,
+            leader_tpu_cache,
+            t_leader_tpu_service,
+        })
+    }
+
+    async fn join(&mut self) {
+        if let Some(t_handle) = self.t_leader_tpu_service.take() {
+            t_handle.await.unwrap().unwrap();
+        }
+    }
+
+    fn leader_tpu_sockets(&self, fanout_slots: u64) -> Vec<SocketAddr> {
+        let current_slot = self.recent_slots.estimated_current_slot();
+        self.leader_tpu_cache
+            .read()
+            .unwrap()
+            .get_leader_sockets(current_slot, fanout_slots)
+    }
+
+    async fn run(
+        rpc_client: Arc<RpcClient>,
+        recent_slots: RecentLeaderSlots,
+        leader_tpu_cache: Arc<RwLock<LeaderTpuCache>>,
+        pubsub_client: Option<PubsubClient>,
+        exit: Arc<AtomicBool>,
+    ) -> Result<()> {
+        let (mut notifications, unsubscribe) = if let Some(pubsub_client) = &pubsub_client {
+            let (notifications, unsubscribe) = pubsub_client.slot_updates_subscribe().await?;
+            (Some(notifications), Some(unsubscribe))
+        } else {
+            (None, None)
+        };
+        let mut last_cluster_refresh = Instant::now();
+        let mut sleep_ms = DEFAULT_MS_PER_SLOT;
+        loop {
+            if exit.load(Ordering::Relaxed) {
+                if let Some(unsubscribe) = unsubscribe {
+                    (unsubscribe)().await;
+                }
+                drop(notifications);
+                if let Some(pubsub_client) = pubsub_client {
+                    pubsub_client.shutdown().await.unwrap();
+                };
+                break;
+            }
+
+            // Sleep a slot before checking if leader cache needs to be refreshed again
+            sleep(Duration::from_millis(sleep_ms)).await;
+            sleep_ms = DEFAULT_MS_PER_SLOT;
+
+            if let Some(notifications) = &mut notifications {
+                while let Ok(Some(update)) =
+                    timeout(Duration::from_millis(10), notifications.next()).await
+                {
+                    let current_slot = match update {
+                        // This update indicates that a full slot was received by the connected
+                        // node so we can stop sending transactions to the leader for that slot
+                        SlotUpdate::Completed { slot, .. } => slot.saturating_add(1),
+                        // This update indicates that we have just received the first shred from
+                        // the leader for this slot and they are probably still accepting transactions.
+                        SlotUpdate::FirstShredReceived { slot, .. } => slot,
+                        _ => continue,
+                    };
+                    recent_slots.record_slot(current_slot);
+                }
+            }
```

Comment:
This is very different. Since the async pubsub API uses fetching from a channel rather than passing in a callbck, we subscribe at the start of the function and poll that on a timeout in this main loop. Any better solutions are welcome!